### ISSUE TITLE
Beam search "cands_to_ignore" is not used correctly.

### ISSUE THREAD

**Issue by jxzhanggg** [facebookresearch/fairseq#3919]
## 🐛 Bug

Cands to ignore is a black list for the beam with EOS. 
The cands to ignore is now used wrongly to compute with the reordered beam, rather than the beam before reorder, i.e. topk operation.

To fix this bug, I think the following line show be deleted:
https://github.com/pytorch/fairseq/blob/f34abcf2b64d8659bae7d3eb2e0caf80cf1d6e7d/fairseq/sequence_generator.py#L408
https://github.com/pytorch/fairseq/blob/f34abcf2b64d8659bae7d3eb2e0caf80cf1d6e7d/fairseq/sequence_generator.py#L491

Then,  after this line:
https://github.com/pytorch/fairseq/blob/f34abcf2b64d8659bae7d3eb2e0caf80cf1d6e7d/fairseq/sequence_generator.py#L537

mask the scores using cands_to_ignore.

**Comment by jxzhanggg** [comment]
The second way to modify the code is to reorder the cands_to_ignore according to the updated beam index, i.e. using the 
"cands_beams".  Current code try to mask using the old beam, but beam order have changed after the topk opertaion.

However, this bug rarely have effect on experiments. In my test set,  the cands_to_ignore is all False. It's hard to trigger 
the cands_to_ignore to play a role in beam search, because it require more than beam size "EOS" appear in top beam size * 2 at one decoding step.

**Comment by mungsoo** [comment]
> The second way to modify the code is to reorder the cands_to_ignore according to the updated beam index, i.e. using the "cands_beams". Current code try to mask using the old beam, but beam order have changed after the topk opertaion.
> 
> However, this bug rarely have effect on experiments. In my test set, the cands_to_ignore is all False. It's hard to trigger the cands_to_ignore to play a role in beam search, because it require more than beam size "EOS" appear in top beam size * 2 at one decoding step.

Hi! Do you think it is possible to have more than beam_size EOS tokens? At any step we only have beam_size * vocab_size lprobs and there are only at most beam_size EOS tokens(at most one EOS for each beam). I'm really confusing why we need cands_to_ignore. Please correct me if I'm wrong. Thank you.

**Comment by yc1999** [comment]
> The second way to modify the code is to reorder the cands_to_ignore according to the updated beam index, i.e. using the "cands_beams". Current code try to mask using the old beam, but beam order have changed after the topk opertaion.
> 
> However, this bug rarely have effect on experiments. In my test set, the cands_to_ignore is all False. It's hard to trigger the cands_to_ignore to play a role in beam search, because it require more than beam size "EOS" appear in top beam size * 2 at one decoding step.

I agree with you. This bug rarely has effect on experiments. In my dataset, `cands_to_ignore` is always full of `False`.
But it is a bug and confuses me.
